src/simulators.jl

```diff
@@ -438,3 +438,110 @@ function B_step!(s, dt_eff, acceleration_vector, compute_forces::Bool, n_threads
     end
     s.velocities += dt_eff * remove_molar.(acceleration_vector)
 end
+
+"""
+    TemperatureREMD(; <keyword arguments>)
+
+A simulator for a parallel temperature replica exchange (TREX) simulation on a [`ReplicaSystem`](@ref). More information on this algorithm can be found in [Sugita Y., Okamoto Y. 1999](https://doi.org/10.1016/S0009-2614(99)01123-9).
+The corresponding [`ReplicaSystem`](@ref) should have the same number of replicas as the number of temperatures in the simulator.
+
+arguments:
+- `dt::S`: the time step of the simulation.
+- `temperatures::TP`: the temperatures corresponding to the replicas.
+- `simulators::ST`: individual simulators for simulating each replica.
+- `exchange_time::ET`: the time interval between replica exchange attempt.
+"""
+struct TemperatureREMD{N, T, S, DT, TP, ST, ET}
+    dt::DT
+    temperatures::TP
+    simulators::ST
+    exchange_time::ET
+end
+
+function TemperatureREMD(;
+                dt,
+                temperatures,
+                simulators,
+                exchange_time,
+                kwargs...)
+    S = eltype(simulators)
+    T = eltype(temperatures)
+    N = length(temperatures)
+    DT = typeof(dt)
+    TP = typeof(temperatures)
+    ET = typeof(exchange_time)
+    if length(simulators) != length(temperatures)
+        throw(ArgumentError("Number of temperatures must match number of simulators"))
+    end
+    if exchange_time <= dt
+        throw(ArgumentError("Exchange time must be greater than the time step"))
+    end
+    simulators = Tuple(simulators[i] for i in 1:N)
+    ST = typeof(simulators)
+    
+    return TemperatureREMD{N, T, S, DT, TP, ST, ET}(temperatures, simulators, exchange_time, dt)
+end
+
+function simulate!(sys::ReplicaSystem,
+                    sim::TemperatureREMD,
+                    n_steps::Int;
+                    n_threads::Int=Threads.nthreads())
+    if sys.n_replicas != length(sim.simulators)
+        error("Number of replicas in ReplicaSystem and simulators in TemperatureREMD must match.")
+    end
+
+    if n_threads > sys.n_replicas
+        thread_div = equal_parts(n_threads, sys.n_replicas)
+    else # pass 1 thread per replica
+        thread_div = equal_parts(sys.n_replicas, sys.n_replicas)
+    end
+
+    # calculate n_cycles and n_steps_per_cycle from dt and exchange_time
+    n_cycles = convert(Int64, (n_steps * sim.dt) ÷ sim.exchange_time)
+    cycle_length = n_steps ÷ n_cycles
+    remaining_steps = n_steps % n_cycles
+
+    # scale to correct temperatures
+    for i in eachindex(sys.replicas)
+        sys.replicas[i].velocities .*= sqrt(sim.temperatures[i] / temperature(sys.replicas[i]))
+    end
```

Comment:
As the same velocities are assigned to each system during replica system setup, I am scaling the velocities to correct temperatures. Here the velocities can also be assigned using `random_velocities!`, but I am not sure which approach is correct from a simulation point of view? Also, should this be optionally controlled by a kwarg in case the simulation is a continuation of some already performed simulation?

Reply:
I think starting velocities should be determined when setting up the `ReplicaSystem`. As you say this works better with continuing simulations, and it is consistent with the other simulators. Perhaps that could take `replica_velocities` instead of `velocities`, and set up the relevant `random_velocities` when `replica_velocities` has its default value of `nothing`.

Reply:
As replica system is a general type so temperature values are not passed to that so we can't use `random_velocities` there. Instead If `replica_velocities` is `nothing`, we can assign velocities to be zero same as `System` type.

Reply:
Ah yes, sounds good. `simulate!` could also have a keyword argument `assign_velocities`, default `false`, that randomizes them in the simulator. Then the user can either set them up manually with the `ReplicaSystem`, giving flexibility and consistency with the other simulators, or call `simulate!` with `assign_velocities=true`.